### PR TITLE
Logging

### PR DESCRIPTION
**Pull request by nrharido**
I implemented logging into the majority of files. I did not convert any of the prints in the UI folder since I was under the impression that they are used in the graphic/text interface (but I can go back and change if I am wrong).

### CODE REVIEW

**Comment by eriq-augustine** [comment]
**Nice!**

I'll let Connor take a pass first, but some notes on the PR (rather than the content of your code):
 1. Make sure to compare against the right branch. (You had `master`, but I just switched it up to `develop`.)
 2. Make sure that commit messages are full sentences and not ambiguous.

For the second point, the exact form of commit messages varies between projects.
But in general, make sure that the message is clear.
A good one for this would be something like: "Implemented an initial logging infrastructure. Replaced non-UI prints with logging statements. Logging is configuration is set in main methods.".
You can take a look at some of our existing commit messages to get a feel for it:
https://github.com/linqs/pacman/commits/develop

File: pacai/agents/learning/reinforcement.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import time
-
+import logging
```
Comment (cfpryor) [inline pacai/agents/learning/reinforcement.py]: import ordering fix

File: pacai/agents/learning/reinforcement.py [hunk 2]
```diff
@@ -17,7 +17,7 @@ class ReinforcementAgent(ValueEstimationAgent):
         - Use self.getLegalActions(state) to know which actions
             are available in a state
     """
-
+    logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.DEBUG)
```
Comment (cfpryor) [inline pacai/agents/learning/reinforcement.py]: Four things:
1) This line should be in the init of this file, we never want lines floating outside functions
2) Why are we setting the DEBUG level here? We want the basic level of logging to be set to INFO and if the students want more output or information they will pass in a flag that is DEBUG (which of course would then change the basic config level to DEBUG.
3) Lastly, we should only be needing to change the debug level in the files that have a main. I.E. any files that sit inside bin (this way we do not need to add this line to every file).
4) We need to have an improved formatting scheme: we want (logginglevel) (time of output) (message), 

I will again say "config issue" for all other cases of this.

File: pacai/agents/search/base.py [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import time
-
+import logging
```
Comment (cfpryor) [inline pacai/agents/search/base.py]: import ordering fix

File: pacai/agents/search/base.py [hunk 4]
```diff
@@ -20,13 +20,13 @@ class SearchAgent(BaseAgent):
     """
 
     # TODO(eriq): We should pass actual objects instead of strings.
-
+    logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.DEBUG)
```
Comment (cfpryor) [inline pacai/agents/search/base.py]: config issue

File: pacai/agents/learning/reinforcement.py [hunk 5]
```diff
@@ -174,24 +174,24 @@ def final(self, state):
...
         if self.episodesSoFar % NUM_EPS_UPDATE == 0:
-                print('Reinforcement Learning Status:')
+                logging.info('Reinforcement Learning Status:')
                 windowAvg = self.lastWindowAccumRewards / float(NUM_EPS_UPDATE)
 
                 if self.episodesSoFar <= self.numTraining:
                     trainAvg = self.accumTrainRewards / float(self.episodesSoFar)
-                    print('\tCompleted %d out of %d training episodes' % (self.episodesSoFar,self.numTraining))
-                    print('\tAverage Rewards over all training: %.2f' % (trainAvg))
+                    logging.info('\tCompleted %d out of %d training episodes' % (self.episodesSoFar,self.numTraining))
```
Comment (cfpryor) [inline pacai/agents/learning/reinforcement.py]: Change these two to debug level logging

File: pacai/agents/learning/reinforcement.py [hunk 6]
```diff
@@ -174,24 +174,24 @@ def final(self, state):
...
+                    logging.info('\tAverage Rewards over all training: %.2f' % (trainAvg))
                 else:
                     testAvg = float(self.accumTestRewards) / (self.episodesSoFar - self.numTraining)
-                    print('\tCompleted %d test episodes' % (self.episodesSoFar - self.numTraining))
-                    print('\tAverage Rewards over testing: %.2f' % (testAvg))
+                    logging.info('\tCompleted %d test episodes' % (self.episodesSoFar - self.numTraining))
```
Comment (cfpryor) [inline pacai/agents/learning/reinforcement.py]: Change these two to debug level logging

File: pacai/bin/capture.py [hunk 7]
```diff
@@ -38,7 +38,7 @@
 
 import pacai.core.layout
 import pacai.util.mazeGenerator
-
+import logging
```
Comment (cfpryor) [inline pacai/bin/capture.py]: import ordering fix

File: pacai/bin/capture.py [hunk 8]
```diff
@@ -832,7 +832,8 @@ def main(argv):
 
     argv already has the executable stripped.
     """
-
+    logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)
+    args = readCommand(argv)  # Get game components based on input
```
Comment (cfpryor) [inline pacai/bin/capture.py]: Duplicate line?

File: pacai/bin/capture.py [hunk 9]
```diff
@@ -832,7 +832,8 @@ def main(argv):
 
     argv already has the executable stripped.
     """
-
+    logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)
```
Comment (cfpryor) [inline pacai/bin/capture.py]: config issue (this is one of the files you want to add it in, but you want to allow it to change here depending on the input from the user).

File: pacai/bin/eightpuzzle.py [hunk 10]
```diff
@@ -1,5 +1,5 @@
 import random
-
+import logging
```
Comment (cfpryor) [inline pacai/bin/eightpuzzle.py]: import fix

File: pacai/bin/eightpuzzle.py [hunk 11]
```diff
@@ -252,18 +252,19 @@ def createRandomEightPuzzle(moves=100):
 
 if __name__ == '__main__':
   puzzle = createRandomEightPuzzle(25)
-  print('A random puzzle:')
-  print(puzzle)
+  logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.DEBUG)
```
Comment (cfpryor) [inline pacai/bin/eightpuzzle.py]: config issue (again right place but needs changing)

File: pacai/bin/eightpuzzle.py [hunk 12]
```diff
@@ -252,18 +252,19 @@ def createRandomEightPuzzle(moves=100):
...
+  logging.info('A random puzzle:')
+  logging.info('\n' + str(puzzle))
```
Comment (cfpryor) [inline pacai/bin/eightpuzzle.py]: Combine to one line

File: pacai/bin/eightpuzzle.py [hunk 13]
```diff
@@ -252,18 +252,19 @@ def createRandomEightPuzzle(moves=100):
...
-  print('BFS found a path of %d moves: %s' % (len(path), str(path)))
+  logging.info('BFS found a path of %d moves: %s' % (len(path), str(path)))
   curr = puzzle
   i = 1
   for a in path:
     curr = curr.result(a)
-    print('After %d move%s: %s' % (i, ("", "s")[i>1], a))
-    print(curr)
+    logging.info('After %d move%s: %s' % (i, ("", "s")[i>1], a))
+    logging.info(curr)
```
Comment (cfpryor) [inline pacai/bin/eightpuzzle.py]: Combine to one line

File: pacai/bin/gridworld.py [hunk 14]
```diff
@@ -1,7 +1,7 @@
 import optparse
 import random
 import sys
-
+import logging
```
Comment (cfpryor) [inline pacai/bin/gridworld.py]: Log order fix

File: pacai/bin/gridworld.py [hunk 15]
```diff
@@ -399,7 +399,7 @@ def runEpisode(agent, environment, discount, decision, display, message, pause,
 
         # EXECUTE ACTION
         nextState, reward = environment.doAction(action)
-        message("Started in state: " + str(state) + "\nTook action: " + str(action) + "\nEnded in state: " + str(nextState) + "\nGot reward: " + str(reward) + "\n")
+        logging.info("\nStarted in state: " + str(state) + "\nTook action: " + str(action) + "\nEnded in state: " + str(nextState) + "\nGot reward: " + str(reward) + "\n")
```
Comment (cfpryor) [inline pacai/bin/gridworld.py]: Change logging level to debug

File: pacai/bin/gridworld.py [hunk 16]
```diff
@@ -478,7 +478,8 @@ def parseOptions():
 
 if __name__ == '__main__':
     opts = parseOptions()
-
+    logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.DEBUG)
```
Comment (cfpryor) [inline pacai/bin/gridworld.py]: config logging

File: pacai/bin/gridworld.py [hunk 17]
```diff
@@ -478,7 +478,8 @@ def parseOptions():
...
+    logging.debug('A random puzzle:\n')
```
Comment (cfpryor) [inline pacai/bin/gridworld.py]: Why is this line in here?

File: pacai/bin/gridworld.py [hunk 18]
```diff
@@ -588,19 +589,13 @@ def update(self, state, action, nextState, reward):
 
     # RUN EPISODES
     if opts.episodes > 0:
-        print()
-        print("RUNNING", opts.episodes, "EPISODES")
-        print()
-
+        logging.info("RUNNING " + str(opts.episodes) + " EPISODES\n")
```
Comment (cfpryor) [inline pacai/bin/gridworld.py]: For now, add in new lines to mimic the old output. Also change to debug

File: pacai/bin/gridworld.py [hunk 19]
```diff
@@ -588,19 +589,13 @@ def update(self, state, action, nextState, reward):
...
     returns = 0
     for episode in range(1, opts.episodes + 1):
         returns += runEpisode(a, env, opts.discount, decisionCallback, displayCallback, messageCallback, pauseCallback, episode)
 
     if opts.episodes > 0:
-        print()
-        print("AVERAGE RETURNS FROM START STATE: " + str((returns + 0.0) / opts.episodes))
-        print()
-        print()
+        logging.info("AVERAGE RETURNS FROM START STATE:" + str((returns + 0.0) / opts.episodes) + "\n")
```
Comment (cfpryor) [inline pacai/bin/gridworld.py]: For now, add in new lines to mimic the old output.  Also change to debug

File: pacai/bin/pacman.py [hunk 20]
```diff
@@ -31,7 +31,7 @@
 import sys
 import time
 import types
-
+import logging
```
Comment (cfpryor) [inline pacai/bin/pacman.py]: Logging ordering

File: pacai/bin/pacman.py [hunk 21]
```diff
@@ -700,7 +700,7 @@ def main(argv):
 
   argv already has the executable stripped.
   """
-
+  logging.basicConfig(format='%(levelname)s: %(message)s', level=logging.DEBUG)
```
Comment (cfpryor) [inline pacai/bin/pacman.py]: Why is basic logging level here DEBUG (it should be INFO), but again should be able to be changed depending on user input.

File: pacai/core/game.py [hunk 22]
```diff
@@ -5,6 +5,7 @@
 import os
 import time
 import traceback
+import logging
```
Comment (cfpryor) [inline pacai/core/game.py]: Logging ordering

File: pacai/core/game.py [hunk 23]
```diff
@@ -390,7 +391,7 @@ def __hash__( self ):
             try:
                 int(hash(state))
             except TypeError as e:
-                print(e)
+                logging.error(e)
```
Comment (cfpryor) [inline pacai/core/game.py]: Add in something before e like logging.error("TypeError %s" % (e))

File: pacai/core/game.py [hunk 24]
```diff
@@ -669,7 +670,7 @@ def run(self):
             except Exception as data:
                 if not self.catchExceptions: raise
                 self.unmute()
-                print("Exception", data)
+                logging.warning("Exception %s", data)
```
Comment (cfpryor) [inline pacai/core/game.py]: Add "%" not ","

File: pacai/core/search/position.py [hunk 25]
```diff
@@ -1,7 +1,7 @@
 from pacai.core.game import Actions
 from pacai.core.game import Directions
 from pacai.core.search.problem import SearchProblem
-
+import logging
```
Comment (cfpryor) [inline pacai/core/search/position.py]: Logging ordering (should be above all this)

File: pacai/core/search/position.py [hunk 26]
```diff
@@ -12,7 +12,7 @@ class PositionSearchProblem(SearchProblem):
 
     Note: this search problem is fully specified; you should NOT change it.
     """
-
+    logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.DEBUG)
```
Comment (cfpryor) [inline pacai/core/search/position.py]: Config logging issue

File: pacai/core/search/position.py [hunk 27]
```diff
@@ -30,7 +30,7 @@ def __init__(self, gameState, costFn = lambda x: 1, goal=(1, 1), start = None, w
         self.costFn = costFn
 
         if (warn and (gameState.getNumFood() != 1 or not gameState.hasFood(*goal))):
-            print('Warning: this does not look like a regular search maze')
+            logging.info('Warning: this does not look like a regular search maze')
```
Comment (cfpryor) [inline pacai/core/search/position.py]: Should be logging.warning

File: pacai/student/analysis.py [hunk 28]
```diff
@@ -4,6 +4,7 @@
 
 # Change these default values to obtain the specified policies through
 # value iteration.
+import logging
```
Comment (cfpryor) [inline pacai/student/analysis.py]: Put above comments

File: pacai/student/analysis.py [hunk 29]
```diff
@@ -131,8 +132,8 @@ def question6():
         question3e,
         question6,
     ]
-
-    print('Answers to analysis questions:')
+    logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.DEBUG)
```
Comment (cfpryor) [inline pacai/student/analysis.py]: Log config issue

File: pacai/student/analysis.py [hunk 30]
```diff
@@ -131,8 +132,8 @@ def question6():
...
+    logging.info('Answers to analysis questions:')
     for question in questions:
         response = question()
-        print('    Question %-10s:\t%s' % (question.__name__, str(response)))
+        logging.info('    Question %-10s:\t%s' % (question.__name__, str(response)))
```
Comment (cfpryor) [inline pacai/student/analysis.py]: Remove spaces before Question

File: pacai/student/searchAgents.py [hunk 31]
```diff
@@ -24,6 +24,7 @@
 from pacai.agents.base import BaseAgent
 from pacai.agents.search.base import SearchAgent
 from pacai.util import util
+import logging
```
Comment (cfpryor) [inline pacai/student/searchAgents.py]: import ordering

File: pacai/student/searchAgents.py [hunk 32]
```diff
@@ -48,7 +48,7 @@ def __init__(self, startingGameState):
         self.corners = ((1, 1), (1, top), (right, 1), (right, top))
         for corner in self.corners:
             if not startingGameState.hasFood(*corner):
-                print('Warning: no food in corner ' + str(corner))
+                logging.info('Warning: no food in corner ' + str(corner))
```
Comment (cfpryor) [inline pacai/student/searchAgents.py]: Change to warning

File: pacai/agents/base.py [hunk 33]
```diff
@@ -95,4 +95,4 @@ def _import_agents(glob_path, package_format_string):
             try:
                 __import__(package_format_string % (module_name))
             except ImportError as ex:
-                print("WARN: Unable to import agent: '%s'. -- %s" % (module_name, str(ex)))
+                logging.warning("Unable to import agent: '%s'. -- %s" % (module_name, str(ex)))
```
Comment (cfpryor) [inline pacai/agents/base.py]: Not going to write this anywhere else, but make sure to use consistent formatting. You switch between double quotes (") and single quotes ('), I think to change the double ones to single.

File: pacai/bin/capture.py [hunk 34]
```diff
@@ -400,10 +400,10 @@ def getProgress(self, game):
 
     def agentCrash(self, game, agentIndex):
         if agentIndex % 2 == 0:
-            print("Red agent crashed")
+            logging.warning("Red agent crashed")
```
Comment (eriq-augustine) [inline pacai/bin/capture.py]: Boost this up to error (along with blue crashes as well).

File: pacai/bin/capture.py [hunk 35]
```diff
@@ -622,7 +622,6 @@ def readCommand( argv ):
     parser.add_option('--gif', dest='gif', help=default('Save the game as a gif to the specified path'))
     parser.add_option('--gif-skip-frames', dest='gifSkipFrames', type='int', default=0, help=default('Skip this number of frames between frames of the gif.'))
     parser.add_option('--gif-fps', dest='gifFPS', type='float', default=10, help=default('FPS of the gif.'))
-
```
Comment (eriq-augustine) [inline pacai/bin/capture.py]: Leave this newline.

File: pacai/bin/capture.py [hunk 36]
```diff
@@ -714,28 +713,28 @@ def randomLayout(seed = None):
     if not seed:
         seed = random.randint(0,99999999)
     # layout = 'layouts/random%08dCapture.lay' % seed
-    # print('Generating random layout in %s' % layout)
+    # logging.info('Generating random layout in %s' % layout)
```
Comment (eriq-augustine) [inline pacai/bin/capture.py]: Remove unused code.

File: pacai/bin/capture.py [hunk 37]
```diff
@@ -714,28 +713,28 @@ def randomLayout(seed = None):
...
+        logging.error('The team "' + agent_module + '" could not be loaded! ')
         traceback.print_exc()
         return [None for i in range(2)]
 
     args = dict()
     args.update(cmdLineArgs) # Add command line args with priority
 
-    print("Loading Team:", agent_module)
-    print("Arguments:", args)
+    logging.debug('Loading Team:%s', agent_module)
```
Comment (eriq-augustine) [inline pacai/bin/capture.py]: Up this one to info.

File: pacai/bin/crawler.py [hunk 38]
```diff
@@ -397,5 +397,6 @@ def _load_args(args):
     return max_steps
 
 if __name__ == '__main__':
+    logging.basicConfig(format='%(levelname)s - %(asctime)s - %(message)s', level=logging.INFO)
```
Comment (eriq-augustine) [inline pacai/bin/crawler.py]: Move all of these calls to a method in a centralized file, how about putting it in `util/logs.py`.
Maybe call the method `initLogging()`?

Note that not calling it `logging` is very purposeful.
We don't want it getting confused with the standard library, `logging`.

File: pacai/bin/eightpuzzle.py [hunk 39]
```diff
@@ -252,18 +252,17 @@ def createRandomEightPuzzle(moves=100):
 
 if __name__ == '__main__':
   puzzle = createRandomEightPuzzle(25)
-  print('A random puzzle:')
-  print(puzzle)
+  logging.basicConfig(format='%(levelname)s - %(asctime)s - %(message)s', level=logging.INFO)
+  logging.info('A random puzzle:\n'+ str(puzzle))
```
Comment (eriq-augustine) [inline pacai/bin/eightpuzzle.py]: Also, almost all binary operators (like `+`) should be surrounded by whitespace.

File: pacai/bin/eightpuzzle.py [hunk 40]
```diff
@@ -252,18 +252,17 @@ def createRandomEightPuzzle(moves=100):
 
 if __name__ == '__main__':
   puzzle = createRandomEightPuzzle(25)
-  print('A random puzzle:')
-  print(puzzle)
+  logging.basicConfig(format='%(levelname)s - %(asctime)s - %(message)s', level=logging.INFO)
```
Comment (eriq-augustine) [inline pacai/bin/eightpuzzle.py]: This should be a print, not a log.
Printing the puzzle like this is part of the core UI in eightpuzzles.

In fact, I think all the logs in this file should go back to prints.
This bin package will probably have the most prints, where as the other packages will most likely be exclusively logs.

File: pacai/bin/gridworld.py [hunk 41]
```diff
@@ -588,23 +587,17 @@ def update(self, state, action, nextState, reward):
 
     # RUN EPISODES
     if opts.episodes > 0:
-        print()
-        print("RUNNING", opts.episodes, "EPISODES")
-        print()
-
+        logging.debug('\nRUNNING ' + str(opts.episodes) + ' EPISODES\n')
```
Comment (eriq-augustine) [inline pacai/bin/gridworld.py]: You can just drop the additional newlines.
(The same below as well.)

File: pacai/bin/pacman.py [hunk 42]
```diff
@@ -294,22 +293,22 @@ def process(self, state, game):
...
+      logging.info('Pacman died! Score: %d' % state.data.score)
     game.gameOver = True
 
   def getProgress(self, game):
     return float(game.state.getNumFood()) / self.initialState.getNumFood()
 
   def agentCrash(self, game, agentIndex):
     if agentIndex == 0:
-      print("Pacman crashed")
+      logging.info('Pacman crashed')
```
Comment (eriq-augustine) [inline pacai/bin/pacman.py]: Raise to error.
(The same for the ghost below.)

File: pacai/student/analysis.py [hunk 43]
```diff
@@ -131,8 +133,7 @@ def question6():
         question3e,
         question6,
     ]
-
-    print('Answers to analysis questions:')
+    logging.info('Answers to analysis questions:')
     for question in questions:
         response = question()
-        print('    Question %-10s:\t%s' % (question.__name__, str(response)))
+        logging.info('Question %-10s:\t%s' % (question.__name__, str(response)))
```
Comment (eriq-augustine) [inline pacai/student/analysis.py]: The log indentation got lost.

File: pacai/bin/eightpuzzle.py [hunk 44]
```diff
@@ -252,8 +253,8 @@ def createRandomEightPuzzle(moves=100):
 
 if __name__ == '__main__':
   puzzle = createRandomEightPuzzle(25)
-  print('A random puzzle:')
-  print(puzzle)
+  initLogging()
```
Comment (eriq-augustine) [inline pacai/bin/eightpuzzle.py]: Make initializing the logging interface the first line in this if statement.

File: pacai/bin/gridworld.py [hunk 45]
```diff
@@ -478,13 +479,13 @@ def parseOptions():
 
 if __name__ == '__main__':
     opts = parseOptions()
-
+    initLogging()
```
Comment (eriq-augustine) [inline pacai/bin/gridworld.py]: Make initializing the logging interface the first line in this if statement.

**Comment by eriq-augustine** [comment]
LGTM

**Comment by eriq-augustine** [comment]
Rebased with a cleaner commit message and merged.
You're in!
ccc139b7232370fb745689b0c4ad4acb11a8231c

**Pull request closed without merging**